packages/flutter/lib/src/widgets/snapshot_widget.dart

```diff
@@ -0,0 +1,269 @@
+  }) : _enabled = enabled;
+
+  /// Reset the snapshot held by any listening [SnapshotWidget].
+  ///
+  /// This has no effect if [enabled] is `false`.
+  void clear() {
+    notifyListeners();
+  }
+
+  /// Whether a snapshot of this child widget is painted in its place.
+  bool get enabled => _enabled;
+  bool _enabled;
+  set enabled(bool value) {
+    if (value == enabled) {
+      return;
+    }
+    _enabled = value;
+    notifyListeners();
+  }
+}
+
+/// A widget that replaces its child with a snapshoted version of the child.
+///
+/// This widget is useful for performing short animations that would otherwise
+/// be expensive or that cannot rely on raster caching. For example, scale and
+/// skew animations are often expensive to perform on complex children, as are
+/// blurs. For a short animation, a widget that contains these expensive effects
+/// can be replaced with a snapshot of it self and manipulated instead.
+///
+/// For example, the Android Q [ZoomPageTransitionsBuilder] uses a snapshot widget
+/// for the forward and entering route to avoid the expensive scale animation.
+/// This also has the effect of briefly pausing any animations on the page.
+///
+/// Generally, this widget should not be used in places where users expect the
+/// child widget to continue animation or to be responsive, such as an unbounded
+/// animation.
+///
+/// Caveats:
+///
+/// * The contents of platform views cannot be captured by a snapshot
+///   widget. If a platform view is encountered, then the snapshot widget will
+///   determine how to render its children based on the [SnapshotMode]. This
+///   defaults to [SnapshotMode.normal] which will throw an exception if a
+///   platform view is encountered.
+///
+/// * This widget is not supported on the HTML backend of Flutter for the web.
+///   On the CanvasKit backend of Flutter, the performance of using this widget
+///   may regress performance due to the fact that both the UI and engine share
+///   a single thread.
+class SnapshotWidget extends SingleChildRenderObjectWidget {
+  /// Create a new [SnapshotWidget].
+  ///
+  /// The [controller] and [child] arguments are required.
+  const SnapshotWidget({
+    super.key,
+    this.mode = SnapshotMode.normal,
+    required this.controller,
+    required super.child
+  });
+
+  /// The controller that determines when to display the children as an snapshot.
+  final SnapshotWidgetController controller;
+
+  /// Configuration that controls how the snapshot widget decides to paint its children.
+  ///
+  /// Defaults to [SnapshotMode.normal], which throws an error when a platform view
+  /// or texture view is encountered.
+  ///
+  /// See [SnapshotMode] for more information.
+  final SnapshotMode mode;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return RenderSnapshotWidget(
+      controller: controller,
+      mode: mode,
+      devicePixelRatio: MediaQuery.maybeOf(context)?.devicePixelRatio ?? 1.0,
```

Comment:
Should we just require that there is a surrounding MediaQuery with `debugCheckHasMediaQuery`? The 1.0 fallback sounds like it could cause some annoying blurriness?